peer/peerstorage.go

```diff
@@ -0,0 +1,103 @@
+package peer
+
+import (
+	"errors"
+
+	"github.com/btcsuite/btcd/btcec/v2"
+	"github.com/lightningnetwork/lnd/kvdb"
+)
+
+var (
+	// peerStorage is the kvdb key where we store other peer's backup data
+	// which they send us.
+	peerStorage = []byte("peer-storage")
+
+	// ErrUninitializedDB is called when the DB is not initialized yet
+	// but a caller calls the `Retrieve` function.
+	ErrUninitializedDB = errors.New("uninitialized Kvdb peer data " +
+		"store")
+)
+
+// KvdbPeerStorage is the kvdb implementation of the PeerDataStore interface.
+type KvdbPeerStorage struct {
+	// DB is th kvdb backend for the storage layer.
+	DB kvdb.Backend
+}
+
+// NewKvdbPeerStorage returns an instance of the kvdb storage layer.
+func NewKvdbPeerStorage(dB kvdb.Backend) (*KvdbPeerStorage, error) {
+	err := kvdb.Update(dB, func(tx kvdb.RwTx) error {
+		_, err := tx.CreateTopLevelBucket(peerStorage)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}, func() {})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &KvdbPeerStorage{
+		DB: dB,
+	}, nil
+}
+
+// Store persists the peer's backup in the storage layer.
+func (k *KvdbPeerStorage) Store(data []byte,
+	peerPub btcec.PublicKey) error {
+
+	return kvdb.Update(k.DB, func(tx kvdb.RwTx) error {
+		bucket, err := tx.CreateTopLevelBucket(peerStorage)
+		if err != nil {
+			return err
+		}
+
+		peerPubBytes := peerPub.SerializeCompressed()
+
+		return bucket.Put(peerPubBytes, data)
+	}, func() {})
+}
+
+// Retrieve fetches the peer's backup from the storage layer.
+func (k *KvdbPeerStorage) Retrieve(peerPub btcec.PublicKey) ([]byte,
+	error) {
+
+	var data []byte
+	err := kvdb.View(k.DB, func(tx kvdb.RTx) error {
+		bucket := tx.ReadBucket(peerStorage)
+
+		peerPubBytes := peerPub.SerializeCompressed()
+
+		if bucket == nil {
+			return ErrUninitializedDB
+		}
```

Comment:
check this directly after the `bucket := tx.ReadBucket(peerStorage)` line

Reply:
I do not understand this one.

Reply:
@ellemouton is saying that the validity check should immediately follow the variable assignment. move these lines up above the pubkey serialization line.

Go relies *heavily* on patterns to enforce things since the language design isn't flexible enough to automate these checks. Deviating from patterns should be done with *extreme* care and in this example it affords no benefit.